### PR TITLE
compiling on msys2 mingw64 gcc5.2.0 and clang

### PR DESCRIPTION
**Pull request by golgepapaz**

### CODE REVIEW

**Comment by narc0tiq** [comment]
Re: your commit dd41694:

> Is there a modern compiler that does not support threads?

It's not about the compiler, but the standard library that comes with it. A quick perusal of Git history shows us this came in around <https://github.com/CleverRaven/Cataclysm-DDA/pull/13152#issuecomment-126948459>. The Jenkins machine builds for Windows using MinGW-w64 via (a fairly old, by now) [mxe](http://mxe.cc/). MSys2 presumably includes the thread support directly.

```
[narc@odin ~]% /home/narc/mxe/usr/bin/i686-w64-mingw32.static-g++ --version 
i686-w64-mingw32.static-g++ (GCC) 4.8.2
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

File: src/vehicle.cpp [hunk 1]
```diff
@@ -3925,91 +3925,6 @@ void vehicle::operate_planter(){
...
-        operate_planter();
-    }
-    if( plow_on ){
-        operate_plow();
-    }
-    if( reaper_on ){
-        operate_reaper();
-    }
-}
-void vehicle::operate_plow(){
```
Comment (BevapDin) [inline src/vehicle.cpp]: You should delete the other versions of the operate_* function. Those here are the good ones that work correctly.

**Comment by golgepapaz** [comment]
I see.Anyway msys2 gcc and stdlib supports threads and inclusion of that file causes duplicate definition errors. I thought this was included to support older mingw compilers. Seeing the msvc project is VS2015 and gcc supports pthreads . the conditional fixes the compile  on this platforms but I can't talk about mxe and jenkins

File: src/vehicle.cpp (hunk 1 above)
Comment (golgepapaz) [inline src/vehicle.cpp]: they look identical to me. are they not?

Comment (golgepapaz) [inline src/vehicle.cpp]: hmm they are not

Comment (golgepapaz) [inline src/vehicle.cpp]: done..

File: src/sounds.cpp [hunk 2]
```diff
@@ -21,7 +21,7 @@
 #ifdef SDL_SOUND
 #include "SDL2/SDL_mixer.h"
 #include <thread>
-#if (defined _WIN32 || defined WINDOWS)
+#if (!defined __MINGW64__ && defined _WIN32 || defined WINDOWS )
```
Comment (BevapDin) [inline src/sounds.cpp]: Is this really what you want or are there some parenthesis missing? Currently it means `(!a && b) || c`, but it should probably be `!a && (b || c)`.

Comment (golgepapaz) [inline src/sounds.cpp]: Yes you are right. Come to think of it is WINDOWS really a macro?. it should be _WINDOWS even than I don't know if that it necessary..

**Comment by narc0tiq** [comment]
For what it's worth, I just did a wintiles build of this very PR and everything worked fine, so you're pretty safe as far as the Jenkins builds are concerned.

File: src/sounds.cpp (hunk 2 above)
Comment (BevapDin) [inline src/sounds.cpp]: As it's already there, leave it there. Never change a running system, but seriously: it's also used in catacurse.h, cata_tiles.h, cursesdef.h and in many more cpp files. It's either correct or at least doesn't cause any trouble.

**Comment by golgepapaz** [comment]
Fixed the if. Narcotiq did you build it with SOUND=1? Thing is only active when sound is enabled.

**Comment by narc0tiq** [comment]
Yeah, SOUND=1 is in <http://dev.narc.ro/cataclysm/mainline-build.zsh>. Will run another test build as soon as I get home and let you know.

**Comment by narc0tiq** [comment]
Latest version built just fine, you're good to go as far as Jenkins is concerned.

**Comment by golgepapaz** [comment]
cool.. now I need to handle the crash in template selection screen.by the way I was able to build this in visual studio because gcc cannot handle the ginormous game.cpp in debug build. MSVC patches need a review though.

**Comment by kevingranade** [comment]
Unfortunately this looks related http://ci.narc.ro/job/Cataclysm-Matrix/Graphics=Tiles,Platform=Windows_x64/3728/console
```
w64-mingw32.static/include/SDL2 -DSDL_SOUND -DLUA -I/home/narc/mxe/usr/x86_64-w64-mingw32.static/include/SDL2 -c src/sounds.cpp -o objwin/tiles/sounds.o
src/sounds.cpp: In function 'void sfx::generate_melee_sound(tripoint, tripoint, bool, bool, std::string)':
src/sounds.cpp:610:5: error: 'thread' is not a member of 'std'
     std::thread the_thread( sound_thread( source, target, hit, targ_mon, material ) );
```

**Comment by golgepapaz** [comment]
Yes. It would seem so .Is it not possible to upgrade the mxe to use something with pthreads support. I am fairly ignorant on the subject though.

**Comment by narc0tiq** [comment]
Funky. The i386 mingw-w64 has it, the x86_64 one doesn't. Let's see if I can do something about that...

**Comment by narc0tiq** [comment]
I don't get it:
```
[narc@odin ~/src/Cataclysm-DDA]% find ~/mxe -iname 'thread'                         
/home/narc/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/4.8.2/include/c++/thread
/home/narc/mxe/usr/lib/gcc/i686-w64-mingw32.static/4.8.2/include/c++/thread
```

The two g++es (`x86_64-w64-mingw32.static-g++` and `i686-w64-mingw32.static-g++`) should be identical, and the `#include <thread>` should've taken care of this whole `'thread' is not a member of 'std'` bullshit.

Maybe I need to update mxe, but as I said at <https://github.com/CleverRaven/Cataclysm-DDA/pull/13152#issuecomment-126948459>, that's both time-consuming and risky.

**Comment by kevingranade** [comment]
For now how about we revert the thing that fixes compilers we don't use as much? (gcc5.2 and clang)

**Comment by narc0tiq** [comment]
Yeah, that sounds like a plan. I'll keep poking at this branch, but there's no reason why it has to be in the blessed master.

**Pull request merged**